File: backend-csharp/Services/RefreshTokenService.cs
using System.Security.Cryptography;
using Microsoft.EntityFrameworkCore;
using Picture2Text.Api.Data;
using Picture2Text.Api.Models;

namespace Picture2Text.Api.Services;

/// <summary>
/// Refresh Token 管理服務
/// </summary>
public class RefreshTokenService
{
    private readonly ApplicationDbContext _context;
    private readonly IConfiguration _configuration;
    private readonly int _refreshTokenExpirationDays;

    public RefreshTokenService(ApplicationDbContext context, IConfiguration configuration)
    {
        _context = context;
        _configuration = configuration;
        _refreshTokenExpirationDays = int.Parse(configuration["Jwt:RefreshTokenExpirationDays"] ?? "7");
    }

    /// <summary>
    /// 生成新的 Refresh Token
    /// </summary>
    public async Task<RefreshToken> GenerateRefreshTokenAsync(
        int userId, 
        string? ipAddress, 
        string? userAgent, 
        string? deviceInfo)
    {
        var tokenId = Guid.NewGuid().ToString();
        var token = GenerateTokenString();

        var refreshToken = new RefreshToken
        {
            UserId = userId,
            Token = HashToken(token), // 儲存加密後的 token
            TokenId = tokenId,
            CreatedAt = DateTime.UtcNow,
            ExpiresAt = DateTime.UtcNow.AddDays(_refreshTokenExpirationDays),
            IsRevoked = false,
            IpAddress = ipAddress,
            UserAgent = userAgent,
            DeviceInfo = deviceInfo
        };

        _context.RefreshTokens.Add(refreshToken);
        await _context.SaveChangesAsync();

        // 返回原始 token（未加密）給客戶端
        refreshToken.Token = token;
        return refreshToken;
    }

    /// <summary>
    /// 驗證 Refresh Token
    /// </summary>
    public async Task<RefreshToken?> ValidateRefreshTokenAsync(string token)
    {
        var hashedToken = HashToken(token);
        
        var refreshToken = await _context.RefreshTokens
            .Include(rt => rt.User)
            .FirstOrDefaultAsync(rt => rt.Token == hashedToken);

        if (refreshToken == null || !refreshToken.IsActive)
        {
            return null;
        }

        // 更新最後使用時間
        refreshToken.LastUsedAt = DateTime.UtcNow;
        await _context.SaveChangesAsync();

        return refreshToken;
    }

    /// <summary>
    /// 撤銷 Refresh Token
    /// </summary>
    public async Task<bool> RevokeRefreshTokenAsync(string token, string? replacedByToken = null)
    {
        var hashedToken = HashToken(token);
        
        var refreshToken = await _context.RefreshTokens
            .FirstOrDefaultAsync(rt => rt.Token == hashedToken);

        if (refreshToken == null || refreshToken.IsRevoked)
        {
            return false;
        }

        refreshToken.IsRevoked = true;
        refreshToken.RevokedAt = DateTime.UtcNow;
        refreshToken.ReplacedByToken = replacedByToken;

        await _context.SaveChangesAsync();
        return true;
    }

    /// <summary>
    /// 撤銷使用者的所有 Refresh Token
    /// </summary>
    public async Task RevokeAllUserTokensAsync(int userId)
    {
        var tokens = await _context.RefreshTokens
            .Where(rt => rt.UserId == userId && !rt.IsRevoked)
            .ToListAsync();

        foreach (var token in tokens)
        {
            token.IsRevoked = true;
            token.RevokedAt = DateTime.UtcNow;
        }

        await _context.SaveChangesAsync();
    }

    /// <summary>
    /// 清理過期的 Refresh Token
    /// </summary>
    public async Task CleanupExpiredTokensAsync()
    {
        var expiredTokens = await _context.RefreshTokens
            .Where(rt => rt.ExpiresAt < DateTime.UtcNow)
            .ToListAsync();

        _context.RefreshTokens.RemoveRange(expiredTokens);
        await _context.SaveChangesAsync();
    }

    /// <summary>
    /// 生成隨機 Token 字串
    /// </summary>
    private static string GenerateTokenString()
    {
        var randomBytes = new byte[64];
        using var rng = RandomNumberGenerator.Create();
        rng.GetBytes(randomBytes);
        return Convert.ToBase64String(randomBytes);
    }

    /// <summary>
    /// 對 Token 進行雜湊加密
    /// </summary>
    private static string HashToken(string token)
    {
        using var sha256 = SHA256.Create();
        var hashBytes = sha256.ComputeHash(System.Text.Encoding.UTF8.GetBytes(token));
        return Convert.ToBase64String(hashBytes);
    }
}

File: backend-nestjs/src/services/refresh-token.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, LessThan } from 'typeorm';
import { ConfigService } from '@nestjs/config';
import * as crypto from 'crypto';
import { RefreshToken } from '../entities';

@Injectable()
export class RefreshTokenService {
  private readonly logger = new Logger(RefreshTokenService.name);
  private readonly refreshTokenExpirationDays: number;

  constructor(
    @InjectRepository(RefreshToken)
    private readonly refreshTokenRepository: Repository<RefreshToken>,
    private readonly configService: ConfigService,
  ) {
    this.refreshTokenExpirationDays = this.configService.get<number>('jwt.refreshTokenExpirationDays')!;
  }

  /**
   * 產生 Refresh Token
   */
  async generateRefreshTokenAsync(
    userId: number,
    tokenId: string,
    ipAddress?: string,
    userAgent?: string,
    deviceInfo?: string,
  ): Promise<{ token: string; expiresAt: Date }> {
    // 產生隨機 Token
    const rawToken = crypto.randomBytes(64).toString('base64url');

    // 計算 Token 的 SHA-256 雜湊（存入資料庫）
    const hashedToken = this.hashToken(rawToken);

    const expiresAt = new Date();
    expiresAt.setDate(expiresAt.getDate() + this.refreshTokenExpirationDays);

    const refreshToken = this.refreshTokenRepository.create({
      UserId: userId,
      Token: hashedToken,
      TokenId: tokenId,
      ExpiresAt: expiresAt,
      IsRevoked: false,
      IpAddress: ipAddress,
      UserAgent: userAgent,
      DeviceInfo: deviceInfo,
    });

    await this.refreshTokenRepository.save(refreshToken);

    this.logger.log(`Generated refresh token for user ${userId}`);

    // 回傳原始 Token（不是雜湊後的）
    return { token: rawToken, expiresAt };
  }

  /**
   * 驗證 Refresh Token
   */
  async validateRefreshTokenAsync(token: string): Promise<RefreshToken | null> {
    const hashedToken = this.hashToken(token);

    const refreshToken = await this.refreshTokenRepository.findOne({
      where: { Token: hashedToken },
      relations: ['User'],
    });

    if (!refreshToken) {
      this.logger.warn('Refresh token not found');
      return null;
    }

    if (refreshToken.IsRevoked) {
      this.logger.warn(`Refresh token ${refreshToken.Id} has been revoked`);
      return null;
    }

    if (new Date() >= refreshToken.ExpiresAt) {
      this.logger.warn(`Refresh token ${refreshToken.Id} has expired`);
      return null;
    }

    // 更新最後使用時間
    refreshToken.LastUsedAt = new Date();
    await this.refreshTokenRepository.save(refreshToken);

    return refreshToken;
  }

  /**
   * 撤銷 Refresh Token
   */
  async revokeRefreshTokenAsync(token: string, replacedByToken?: string): Promise<boolean> {
    const hashedToken = this.hashToken(token);

    const refreshToken = await this.refreshTokenRepository.findOne({
      where: { Token: hashedToken },
    });

    if (!refreshToken) {
      return false;
    }

    refreshToken.IsRevoked = true;
    refreshToken.RevokedAt = new Date();
    if (replacedByToken) {
      refreshToken.ReplacedByToken = this.hashToken(replacedByToken);
    }

    await this.refreshTokenRepository.save(refreshToken);

    this.logger.log(`Revoked refresh token ${refreshToken.Id}`);

    return true;
  }

  /**
   * 撤銷使用者的所有 Refresh Token
   */
  async revokeAllUserTokensAsync(userId: number): Promise<number> {
    const result = await this.refreshTokenRepository.update(
      { UserId: userId, IsRevoked: false },
      { IsRevoked: true, RevokedAt: new Date() },
    );

    const count = result.affected || 0;
    this.logger.log(`Revoked ${count} refresh tokens for user ${userId}`);

    return count;
  }

  /**
   * 根據 Token ID 取得 Refresh Token
   */
  async getByTokenIdAsync(tokenId: string): Promise<RefreshToken | null> {
    return this.refreshTokenRepository.findOne({
      where: { TokenId: tokenId },
      relations: ['User'],
    });
  }

  /**
   * 清理過期的 Refresh Token
   */
  async cleanupExpiredTokensAsync(): Promise<number> {
    const result = await this.refreshTokenRepository.delete({
      ExpiresAt: LessThan(new Date()),
    });

    const count = result.affected || 0;
    if (count > 0) {
      this.logger.log(`Cleaned up ${count} expired refresh tokens`);
    }

    return count;
  }

  /**
   * 計算 Token 的 SHA-256 雜湊
   */
  private hashToken(token: string): string {
    return crypto.createHash('sha256').update(token).digest('hex');
  }
}
